src/mir-regionck.md

```diff
@@ -0,0 +1,376 @@
+sometimes be smarter here (and in fact we have code for doing this
+already in other contexts). Moreover, since `'static` is in U0, we
+know that all variables can see it -- so basically if we find a that
+`value(V2)` contains `skol(u)` for some universe `u` that `V1` can't
+see, then we force `V1` to `'static`.
+
+### Extending the "universal regions" check
+
+After all constraints have been propagated, the NLL region inference
+has one final check, where it goes over the values that wound up being
+computed for each universal region and checks that they did not get
+'too large'. In our case, we will go through each skolemized region
+and check that it contains *only* the `skol(u)` element it is known to
+outlive. (Later, we might be able to know that there are relationships
+between two skolemized regions and take those into account, as we do
+for universal regions from the fn signature.)
+
+Put another way, the "universal regions" check can be considered to be
+checking constraints like:
+
+    {skol(1)}: V1
+    
+where `{skol(1)}` is like a constant set, and V1 is the variable we
+made to represent the `!1` region.
+
+## Back to our example
+
+OK, so far so good. Now let's walk through what would happen with our
+first example:
+
+    fn(&'static u32) <: fn(&'!1 u32) @ P  // this point P is not imp't here
+
+The region inference engine will create a region element domain like this:
+
+    { CFG; end('static); skol(1) }
+      ---  ------------  ------- from the universe `!1`
+      |    'static is always in scope
+      all points in the CFG; not especially relevant here 
+
+It will always create two universal variables, one representing
+`'static` and one representing `'!1`. Let's call them Vs and V1. They
+will have initial values like so:
+
+    Vs = { CFG; end('static) } // it is in U0, so can't name anything else
+    V1 = { skol(1) }
+    
+From the subtyping constraint above, we would have an outlives constraint like
+
+    '!1: 'static @ P
+
+To process this, we would grow the value of V1 to include all of Vs:
+
+    Vs = { CFG; end('static) }
+    V1 = { CFG; end('static), skol(1) }
+
+At that point, constraint propagation is done, because all the
+outlives relationships are satisfied. Then we would go to the "check
+universal regions" portion of the code, which would test that no
+universal region grew too large.
+
+In this case, `V1` *did* grow too large -- it is not known to outlive
+`end('static)`, nor any of the CFG -- so we would report an error.
+
+## Another example
+
+What about this subtyping relationship?
+
+    for<'a> fn(&'a u32, &'a u32)
+        <:
+    for<'b, 'c> fn(&'b u32, &'c u32)
+    
+Here we would skolemize the supertype, as before, yielding:    
+
+    for<'a> fn(&'a u32, &'a u32)
+        <:
+    fn(&'!1 u32, &'!2 u32)
+    
+then we instantiate the variable on the left-hand side with an existential
+in universe U2, yielding:
+
+    fn(&'?3 u32, &'?3 u32) 
```

Comment:
I think this is the first time '?n is introduced as notation for existential n...

Reply:
added a comment